src/libopensc/card-piv.c

```diff
@@ -2965,7 +3213,47 @@ static int piv_pin_cmd(sc_card_t *card, struct sc_pin_cmd_data *data,
 			}
 		}
 	}
-	return iso_drv->ops->pin_cmd(card, data, tries_left);
+
+	priv->pin_cmd_verify_sw1 = 0x00U;
+
+	priv->pin_cmd_verify = 1; /* tell piv_check_sw its a verify to save sw1, sw2 */
+	r = iso_drv->ops->pin_cmd(card, data, tries_left);
+	priv->pin_cmd_verify = 0;
+
+	/* if access to applet is know to be reset by other driver  we select_aid and try again */
+	if ( priv->card_issues & CI_OTHER_AID_LOSE_STATE && priv->pin_cmd_verify_sw1 == 0x6DU) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "AID may be lost doing piv_find_aid and retry pin_cmd");
+		r = piv_find_aid(card, priv->aid_file); /* return not tested */
+
+		priv->pin_cmd_verify = 1; /* tell piv_check_sw its a verify to save sw1, sw2 */
+		r = iso_drv->ops->pin_cmd(card, data, tries_left);
+		priv->pin_cmd_verify = 0;
+	}
+
+	/* Some cards never return 90 00  for SC_PIN_CMD_GET_INFO even if the card state is verified */
+
+	if (data->cmd == SC_PIN_CMD_GET_INFO
+			&& priv->card_issues & CI_VERIFY_LC0_FAIL
+			&& priv->pin_cmd_verify_sw1 == 0x63U /* can not use SC_ERROR_PIN_CODE_INCORRECT here */
+			&& !(priv->card_issues & CI_CANT_USE_GETDATA_FOR_STATE)) {
+				/* try another method, looking at a protected object */
+				r = piv_check_protected_objects(card);
+	}
+
+	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
+}
+
+
+static int piv_logout(sc_card_t *card)
+{
+	int r = 0;
+	/* piv_private_data_t * priv = PIV_DATA(card); */
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	/* TODO 800-73-3 does not define a logout, 800-73-4 does */
+
+	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
```

Comment:
If the card doesn't implement a logout function, you should return `SC_ERROR_NOT_SUPPORTED` instead of ignoring the explicit logout. If a logout according to 800-73-3 (whatever this may be) is not possible you could use `sc_reset` instead.